compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Fetch.hs

```diff
@@ -0,0 +1,100 @@
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+
+import DA.Daml.Compiler.Dar (PackageSdkVersion(..),createDarFile,createArchive)
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Ast.Optics as LF (packageRefs)
+import qualified DA.Daml.LF.Proto3.Archive as LFArchive
+import qualified DA.Ledger as L
+import qualified SdkVersion
+
+data LedgerArgs = LedgerArgs
+  { host :: String
+  , port :: Int
+  , tokM :: Maybe L.Token
+  , sslConfigM :: Maybe L.ClientSSLConfig
+  }
+
+instance Show LedgerArgs where
+  show LedgerArgs{host,port} = host <> ":" <> show port
+
+-- | Reconstruct a DAR file by downloading packages from a ledger. Returns how many packages fetched.
+fetchDar :: LedgerArgs -> LF.PackageId -> FilePath -> IO Int
+fetchDar ledgerArgs rootPid saveAs = do
+  xs <- downloadAllReachablePackages ledgerArgs rootPid
+  [pkg] <- pure [ pkg | (pid,pkg) <- xs, pid == rootPid ]
+  let (dalf,pkgId) = LFArchive.encodeArchiveAndHash pkg
+  let dalfDependencies :: [(T.Text,BS.ByteString,LF.PackageId)] =
+        [ (txt,bs,pkgId)
+        | (pid,pkg) <- xs, pid /= rootPid
+        , let txt = T.pack ("dep-" <> T.unpack (LF.unPackageId pid))
+        , let (bsl,pkgId) = LFArchive.encodeArchiveAndHash pkg
+        , let bs = BSL.toStrict bsl
+        ]
+  let pName :: LF.PackageName = LF.PackageName $ T.pack "reconstructed"
+  let pVersion = Nothing
+  let pSdkVersion = PackageSdkVersion SdkVersion.sdkVersion
+  let srcRoot = error "unexpected use of srcRoot when there are no sources"
+  let za = createArchive pName pVersion pSdkVersion pkgId dalf dalfDependencies srcRoot [] [] []
+  createDarFile saveAs za
+  return $ length xs
+
+-- | Download all Packages reachable from a PackageId; fail if any don't exist or can't be decoded.
+downloadAllReachablePackages :: LedgerArgs -> LF.PackageId -> IO [(LF.PackageId,LF.Package)]
+downloadAllReachablePackages ledgerArgs pid = loop [] [pid]
+  where
+    loop :: [(LF.PackageId,LF.Package)] -> [LF.PackageId] -> IO [(LF.PackageId,LF.Package)]
+    loop acc = \case
+      [] -> return acc
+      pid:morePids ->
+        if pid `elem` [ pid | (pid,_) <- acc ]
+        then loop acc morePids
+        else do
+          pkg <- downloadPackage ledgerArgs pid
+          loop ((pid,pkg):acc) (packageRefs pkg ++ morePids)
+
+packageRefs :: LF.Package -> [LF.PackageId]
+packageRefs pkg = nubSort [ pid | LF.PRImport pid <- toListOf LF.packageRefs pkg ]
+
+-- | Download the Package identified by a PackageId; fail if it doesn't exist or can't be decoded.
+downloadPackage :: LedgerArgs -> LF.PackageId -> IO LF.Package
+downloadPackage ledgerArgs pid = do
+  let ls :: L.LedgerService (Maybe L.Package) = do
+        lid <- L.getLedgerIdentity
+        L.getPackage lid $ convPid pid
+  runWithLedgerArgs ledgerArgs ls >>= \case
+    Nothing -> error $ "Unable to download package with identity: " <> show pid
```

Comment:
```suggestion
    Nothing -> fail $ "Unable to download package with identity: " <> show pid
```
Should be fine here but in general `error` in IO is a bit fishy since it throws when evaluated rather than when executed.

Reply:
done